osu.Game/Database/ArchiveModelManager.cs

```diff
@@ -0,0 +1,337 @@
+                    // Also, not always a single file, i.e. for LegacyFilesystemReader
+                    // TODO: Add a check to prevent files from storage to be deleted.
+                    try
+                    {
+                        if (File.Exists(path))
+                            File.Delete(path);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Error(e, $@"Could not delete original file after import ({Path.GetFileName(path)})");
+                    }
+                }
+                catch (Exception e)
+                {
+                    e = e.InnerException ?? e;
+                    Logger.Error(e, $@"Could not import ({Path.GetFileName(path)})");
+                }
+            }
+
+            notification.State = ProgressNotificationState.Completed;
+        }
+
+        /// <summary>
+        /// Import an item from an <see cref="ArchiveReader"/>.
+        /// </summary>
+        /// <param name="archive">The archive to be imported.</param>
+        public TModel Import(ArchiveReader archive)
+        {
+            using (ContextFactory.GetForWrite()) // used to share a context for full import. keep in mind this will block all writes.
+            {
+                // create a new model (don't yet add to database)
+                var item = CreateModel(archive);
+
+                var existing = CheckForExisting(item);
+
+                if (existing != null) return existing;
+
+                item.Files = createFileInfos(archive, Files);
+
+                Populate(item, archive);
+
+                // import to store
+                ModelStore.Add(item);
+
+                return item;
+            }
+        }
+
+        /// <summary>
+        /// Import an item from a <see cref="TModel"/>.
+        /// </summary>
+        /// <param name="item">The model to be imported.</param>
+        public void Import(TModel item) => ModelStore.Add(item);
+
+        /// <summary>
+        /// Perform an update of the specified item.
+        /// TODO: Support file changes.
+        /// </summary>
+        /// <param name="item">The item to update.</param>
+        public void Update(TModel item) => ModelStore.Update(item);
+
+        /// <summary>
+        /// Delete an item from the manager.
+        /// Is a no-op for already deleted items.
+        /// </summary>
+        /// <param name="item">The item to delete.</param>
+        public void Delete(TModel item)
+        {
+            using (var usage = ContextFactory.GetForWrite())
+            {
+                var context = usage.Context;
+
+                context.ChangeTracker.AutoDetectChangesEnabled = false;
+
+                // re-fetch the model on the import context.
+                var foundModel = queryModel().Include(s => s.Files).ThenInclude(f => f.FileInfo).First(s => s.ID == item.ID);

```

Comment:
Is this going to be safe? It looks like `queryModel()` always returns read-only contexts, which do not track whether they've written.

Reply:
I guess it is since ModelStore.Delete is run on the write context?

Reply:
We're inside a write context. Remember the write context is just using the current thread's read context, so this is always the same context.

Reply:
The second part of that was the more critical point - non-write context don't track whether they've written anything.

Reply:
I'm not sure what you mean by that.

Reply:
See line 176 – we're already within a write usage.

Reply:
Right, I see.